flang/lib/Optimizer/CodeGen/TBAABuilder.cpp

```diff
@@ -39,26 +49,10 @@ static llvm::cl::opt<unsigned>
 namespace fir {
 
 TBAABuilder::TBAABuilder(MLIRContext *context, bool applyTBAA)
-    : enableTBAA(applyTBAA && !disableTBAA) {
+    : enableTBAA(applyTBAA && !disableTBAA),
+      trees(/*separatePerFunction=*/perFunctionTBAATrees) {
```

Comment:
Do I understand it correctly that even without running the FIR TBAA attachment pass we will use per-function trees by default after this change?  I wonder if we can still stick to a single global tree for those functions that have not been affected by the FIR TBAA attachment pass?

Reply:
Yes we will. Although, my intention is to enable my FIR pass whenever we are using TBAA so I don't anticipate TBAABuilder will often run without the FIR alias tags pass..

I am yet to see evidence of significant slowdowns due to using per-function TBAA trees. The FIR pass applies tags to every function which has dummy arguments, so I suspect keeping the global tree for unaffected functions which slow down due to per-function trees is a very uncommon case.

Reply:
Thanks for explaining.  My only concern is a "clean" switch back to the previous state of things (e.g. to triage new failures, if any).  With this change, we will have to specify two options to switch to the TBAA that we currently have, i.e. we will have to disable the new FIR TBAA pass, and also disable the per-function behavior for the codegen.  It is not a big deal, I just wanted to confirm that this is the case.

Reply:
I think so, but I am okay with having the two flags as well.

Reply:
Done. See https://github.com/llvm/llvm-project/pull/68437/commits/56ed3a6e25c34542a47d52afc7cf1aeaa2b58271 and https://github.com/llvm/llvm-project/pull/68595/commits/9168aa0f6722d576267747e3ec3f8178e6df7f85

Reply:
Thank you! It looks good to me.